### PR TITLE
Resprites SMES, Port from Haven-Urist

### PR DESCRIPTION
**Pull request by Erol509**
## About The Pull Request

New SMES sprites by Irra or @martinlyra from https://github.com/Haven-13/Haven-Urist/pull/67
SMES's have a EMP interaction now.

image
image

## Why It's Good For The Game

New SMES sprites, it still has similar SOUL but has more modern display, I think it looks nice, and fits quite nice. 

## Changelog

:cl:
add: New SMES sprites made by Irra and retlaw34 for panel icon. Credits for them. 
/:cl:

### CODE REVIEW

**Comment by ZephyrTFA** [comment]
please include side by side comparisons for all states of the SMES

**Comment by Erol509** [comment]
> please include side by side comparisons for all states of the SMES

For all the overlays? It would be hard to get that in game

**Comment by ZephyrTFA** [comment]
yes

**Comment by ZephyrTFA** [comment]
you can just do the easy ones,
low power
medium power
fully charged
disabled

**Comment by Erol509** [comment]
> you can just do the easy ones, low power medium power fully charged disabled

image
Top one fully OFF, second one (full power charge) and sending indicator. 
image
without sending indicator
image
0% charge, charging on auto.

**Comment by Erol509** [comment]
image
30% charge, the bar goes up. Red dot near the charge bar means, not charging, yellow means charging. And when it has 2 green bars on top it means its fully powered

**Comment by ZephyrTFA** [comment]
I like the sprites but let's wait for spritetainer input

**Comment by Erol509** [comment]
128158604-1cb17b84-996d-41bb-a5b5-c121a1171777

File: code/modules/power/smes.dm [hunk 1]
```diff
@@ -221,18 +227,60 @@
...
+		return
 
 	var/clevel = chargedisplay()
-	if(clevel > 0)
-		. += "smes-og[clevel]"
+	if(!terminal)
+		. += "failing"
+		SSvis_overlays.add_vis_overlay(src, icon, "failing", layer, plane, dir)
+		SSvis_overlays.add_vis_overlay(src, icon, "failing", layer, EMISSIVE_PLANE, dir)
+	if(clevel>0)
```
Comment (LemonInTheDark) [inline code/modules/power/smes.dm:244]: spaces around operators please

File: code/modules/power/smes.dm [hunk 2]
```diff
@@ -221,18 +227,60 @@
...
+				SSvis_overlays.add_vis_overlay(src, icon, "output2", layer, plane, dir)
+				SSvis_overlays.add_vis_overlay(src, icon, "output2", layer, EMISSIVE_PLANE, dir)
+			else
+				. += "output1"
+				SSvis_overlays.add_vis_overlay(src, icon, "output1", layer, plane, dir)
+				SSvis_overlays.add_vis_overlay(src, icon, "output1", layer, EMISSIVE_PLANE, dir)
+		else
+			. += "output0"
+			SSvis_overlays.add_vis_overlay(src, icon, "output0", layer, plane, dir)
+			SSvis_overlays.add_vis_overlay(src, icon, "output0", layer, EMISSIVE_PLANE, dir)
```
Comment (LemonInTheDark) [inline code/modules/power/smes.dm:283]: use early returns <https://github.com/tgstation/tgstation/blob/master/.github/guides/STYLE.md#use-early-returns> at least so far as is possible (likely will only work with the uppermost block)

File: code/modules/power/smes.dm [hunk 3]
```diff
@@ -435,6 +485,12 @@
 	charge = INFINITY
 	..()
 
+/obj/machinery/power/smes/proc/emp_end() //used to check if SMES was EMPED and reset the overlay state
+	is_emped = FALSE
+	update_icon()
+	log_smes()
+
```
Comment (LemonInTheDark) [inline code/modules/power/smes.dm:492]: one less newline please, also what is this log FOR if being emp'd isn't a part of its readout

File: code/modules/power/smes.dm [hunk 4]
```diff
@@ -221,18 +227,60 @@
...
+	if(machine_stat & BROKEN)
+		return
 
 	var/clevel = chargedisplay()
-	if(clevel > 0)
-		. += "smes-og[clevel]"
+	if(!terminal)
+		. += "failing"
+		SSvis_overlays.add_vis_overlay(src, icon, "failing", layer, plane, dir)
+		SSvis_overlays.add_vis_overlay(src, icon, "failing", layer, EMISSIVE_PLANE, dir)
```
Comment (LemonInTheDark) [inline code/modules/power/smes.dm:243]: the vis_overlays system is deprecated, has been for years (vis_contents, which is how it works, create higher server load so we do not use them)

Instead, use `mutable_appearance` and `emissive_appearance` (alongside their arguments)
Also, `. += "{icon_state}"` is handled the same as just handing back a mutable appearance with the particular icon state set and all other values inherited from the parent. So these states are being displayed twice incorrectly.

Also of note, emissive_appearance takes an atom to use as a reference for its position in the map (to make multiz work correctly). The smes will likely work just fine for this

File: code/modules/power/smes.dm (hunk 2 above)
Comment (LemonInTheDark) [inline code/modules/power/smes.dm:283]: This seems a little dumb. Might be best to work out the clevel number first, and then reuse it wherever it is called for.
Also of note, why do the input icon states separate with a - and the output ones don't? that seems dumb

File: code/modules/power/smes.dm [hunk 5]
```diff
@@ -410,6 +458,8 @@
 	. = ..()
 	if(. & EMP_PROTECT_SELF)
 		return
+	emp_timer = addtimer(CALLBACK(src, .proc/emp_end, output_attempt), SMESEMPTIME, TIMER_UNIQUE | TIMER_OVERRIDE)
```
Comment (LemonInTheDark) [inline code/modules/power/smes.dm:461]: you need to use PROC_REF here, since we run on 515

File: code/modules/power/smes.dm [hunk 6]
```diff
@@ -14,9 +14,12 @@
 #define SMES_INPUTTING 8
 #define SMES_INPUT_ATTEMPT 9
 
+#define SMESEMPTIME 20 SECONDS // the time it takes for the SMES to go back to normal operation when emped
```
Comment (LemonInTheDark) [inline code/modules/power/smes.dm:17]: why is this a define? it's only used once?
Also, if it remains one please format it consistently with the others

File: code/modules/power/smes.dm [hunk 7]
```diff
@@ -221,18 +227,60 @@
 
 /obj/machinery/power/smes/update_overlays()
 	. = ..()
-	if(machine_stat & BROKEN)
-		return
+	SSvis_overlays.remove_vis_overlay(src, managed_vis_overlays)
```
Comment (LemonInTheDark) [inline code/modules/power/smes.dm:230]: this code becomes redundant if vis_overlays is nuked

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]
Code review, much of this is born from you porting directly from haven code, should be careful when doin that in future as you learn more, we have different standards and systems and all. Not a huge deal tho

Pretty sprites (I think idk they don't pay me to art)

**Comment by Wallemations** [comment]
Why did you make a new .dmi and not just edit the previously existing SMES icon states?

File: code/modules/power/smes.dm (hunk 7 above)
Comment (Erol509) [reply code/modules/power/smes.dm:230]: I have no idea how to make it other way.

**Comment by Erol509** [comment]
> Why did you make a new .dmi and not just edit the previously existing SMES icon states?

Honestly? I couldnt find it, and smes.dm didnt point to the exact location.

**Comment by Erol509** [comment]
I will apply changes tomorrow.

**Comment by Wallemations** [comment]
> > Why did you make a new .dmi and not just edit the previously existing SMES icon states?
> 
> 
> 
> Honestly? I couldnt find it, and smes.dm didnt point to the exact location. 

Then look further up the typetree

**Comment by MTandi** [comment]
I dislike it because I don't understand the shape by looking at the sprite. 
The current SMES sprite is dated, but it reads clearly and subjectively better than this version.
Also light blue is bad for the indicator because it is close to the main color of the SMES.

File: code/modules/power/smes.dm (hunk 2 above)
Comment (Erol509) [reply code/modules/power/smes.dm]: Anywhere else than under EMP state, it wont work and block overlays.

Comment (Erol509) [reply code/modules/power/smes.dm]: Tried to redo using if(inputting && clevel == SMES_CLEVEL_5) but it doesnt seem to work as intended after testing it. Also for the other thing, it was just typos under dmi states names, fixed them to have output-

**Comment by martinlyra** [comment]
Hi yes. I figured I would go add some context to here. This PR seems to similar to everything about this original port PR to shiptest here: https://github.com/shiptest-ss13/Shiptest/pull/1099

And as for the sprite design. I didn't get much feedback regarding the sprites since the sample size was small. There were two things I had in mind when I made the sprites:
- The original sprite on Baystation was not pleasant to look at because the indicators have the same contrast as the body, with no outlines to make them more visible. The front panel felt too small to add major black outlines to make the indicators pop out more.
- Baystation code had no emissive system at the time, the only way to read them was to open their UI. It getting emissives was an afterthought. But when it came to Haven, it was intentional to make most of the game dimly-lit with half the areas fully lit. SMES were usually in the dark corners of the map so they popped out "fine" at glance.
I am glad to see a new perspective and feedback.

Oh by the way. The 'broken' icon-state was a relic of Baystation code that Haven is based on. Meant to be immediate danger before exploding. Which happened when shot or heavily damaged while at least near full charge if remember the old code correctly. If you don't need it because you don't have this feature, then you most likely do not need it. It was supposed to be completely dark when broke back in Haven.

Also it was @retlaw34's work doing the PCB sprite when opened (added in same PR I linked earlier https://github.com/shiptest-ss13/Shiptest/pull/1099), so, please; credit her too.

**Comment by Erol509** [comment]
> Hi yes. I figured I would go add some context to here. This PR seems to similar to everything about this original port PR to shiptest here: [shiptest-ss13/Shiptest#1099](https://github.com/shiptest-ss13/Shiptest/pull/1099)
> 
> And as for the sprite design. I didn't get much feedback regarding the sprites since the sample size was small. There were two things I had in mind when I made the sprites:
> 
> * The original sprite on Baystation was not pleasant to look at because the indicators have the same contrast as the body, with no outlines to make them more visible. The front panel felt too small to add major black outlines to make the indicators pop out more.
> * Baystation code had no emissive system at the time, the only way to read them was to open their UI. It getting emissives was an afterthought. But when it came to Haven, it was intentional to make most of the game dimly-lit with half the areas fully lit. SMES were usually in the dark corners of the map so they popped out "fine" at glance.
> 
> I am glad to see a new perspective and feedback.
> 
> Oh by the way. The 'broken' icon-state was a relic of Baystation code that Haven is based on. Meant to be immediate danger before exploding. Which happened when shot or heavily damaged while at least near full charge if remember the old code correctly. If you don't need it because you don't have this feature, then you most likely do not need it. It was supposed to be completely dark when broke back in Haven.
> 
> Also it was @retlaw34's work doing the PCB sprite when opened (added in same PR I linked earlier [shiptest-ss13/Shiptest#1099](https://github.com/shiptest-ss13/Shiptest/pull/1099)), so, please; credit her too.

Indeed I took it from shiptest, and sure for the credits

**Review by Wallemations — CHANGES_REQUESTED** [review]
I'm struggling to put my thoughts into words but I think the biggest gripe I have on this one is the lower visual acuity in respects to the general shape of the machine.
Next to a lot of the other 3/4ths sprites we have it feels a little thin/flat, and it loses out on the previous hexagonal design that made the previous sprite recognizable

I do like the shading though maybe a little more saturation could help.
I think really the issue could be solved by fixing the perspective, in this case focusing on the top of the machine and emphasizing its hexagonal shape.

**Pull request closed without merging**